xarray/core/indexes.py

```diff
@@ -138,64 +134,65 @@ def get_indexer_nd(index, labels, method=None, tolerance=None):
     return indexer
 
 
-class PandasIndex(Index, ExplicitlyIndexedNDArrayMixin):
-    """Wrap a pandas.Index to preserve dtypes and handle explicit indexing."""
+def _create_variables_from_index(index, dim, attrs=None, encoding=None):
+    from .variable import IndexVariable
 
-    __slots__ = ("array", "_dtype")
+    if index.name is None:
+        name = dim
+    else:
+        name = index.name
 
-    def __init__(
-        self, array: Any, dtype: DTypeLike = None, coord_name: Optional[Hashable] = None
-    ):
-        if coord_name is None:
-            coord_name = tuple()
-        super().__init__(coord_name)
+    data = LazilyIndexedArray(PandasIndexingAdapter(index))
+    var = IndexVariable(dim, data, attrs=attrs, encoding=encoding, fastpath=True)
 
-        self.array = utils.safe_cast_to_index(array)
+    return {name: var}
 
-        if dtype is None:
-            if isinstance(array, pd.PeriodIndex):
-                dtype_ = np.dtype("O")
-            elif hasattr(array, "categories"):
-                # category isn't a real numpy dtype
-                dtype_ = array.categories.dtype
-            elif not utils.is_valid_numpy_dtype(array.dtype):
-                dtype_ = np.dtype("O")
-            else:
-                dtype_ = array.dtype
+
+class PandasIndex(Index):
+    """Wrap a pandas.Index as an xarray compatible index."""
+
+    __slots__ = ("index", "coords")
+
+    def __init__(self, variables, fastpath=False):
+        if len(variables) != 1:
+            raise ValueError(
+                f"PandasIndex only accepts one variable, found {len(variables)} variables"
+            )
+
+        name, var = next(iter(variables.items()))
+
+        if var.ndim != 1:
+            raise ValueError(
+                "PandasIndex only accepts a 1-dimensional variable, "
+                f"variable {name!r} has {var.ndim} dimensions"
+            )
+
+        if fastpath:
+            self.index = var.data.array
+            self.coords = {name: var}
         else:
-            dtype_ = np.dtype(dtype)  # type: ignore[assignment]
-        self._dtype = dtype_
+            self.index = utils.safe_cast_to_index(var.data)
+            self.index.name = name
+            self.coords = _create_variables_from_index(
+                self.index, var.dims[0], attrs=var.attrs, encoding=var.encoding
+            )
```

Comment:
Note: This version means you *always* need to override the variables with `index.coords`. A separate class method constructor would in principle provide more flexibility, by returning a separate dict of `variables`.

Reply:
I guess that in most cases we'll need to override the variables, at least convert the `Variable` objects into `IndexVariable` object to ensure immutability. Probably the only case were we don't need to override them is when we already have an index object (e.g., a new `pd.Index` as result from a selection / alignment / etc or if we still allow passing it to `DataArray` / `Dataset` constructors). But in the latter case we still need to create the variables from the index, which is better achieved using (another) class method constructor.

In general, I agree that a class method constructor returning a separate dict of variables is probably better also for the former cases. This provides clearer API, which ultimately will be public.

We could then get rid of the `coords` attribute, with a (minor) concern that we might have to pass the overridden variables through several (many?) functions.